cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,489 @@
+                         cudf::column_view const& right_col,
+                         cudf::mutable_column_view& out_col,
+                         index_type const* merged_indices,
+                         cudaStream_t stream) {
+  constexpr cudf::size_type BLOCK_SIZE{256};
+  cudf::experimental::detail::grid_1d grid_config {out_col.size(), BLOCK_SIZE };
+
+  auto p_left_dcol  = cudf::column_device_view::create(left_col);
+  auto p_right_dcol = cudf::column_device_view::create(right_col);
+  auto p_out_dcol   = cudf::mutable_column_device_view::create(out_col);
+
+  auto left_valid  = *p_left_dcol;
+  auto right_valid = *p_right_dcol;
+  auto out_valid   = *p_out_dcol;
+
+  //these tests in the legacy code
+  //tested the null_mask buffer against nullptr,
+  //not if there were nulls, which may not be
+  //equivalent with semantics below...
+  //
+  //in fact, the null_mask being nullptr is
+  //equivalent to ALL_VALID (see types.hpp comment on
+  //UNALLOCATED);
+  //this is not the meaning of
+  //left_have_valids and right_have_valids non-template
+  //bool params (which indicate whether the corresponding
+  //null_maks buffers are non-nullptr<true> or not<false>);
+  //
+  //
+  if (p_left_dcol->has_nulls()) {
+    if (p_right_dcol->has_nulls()) {
+      materialize_merged_bitmask_kernel<true, true>
+        <<<grid_config.num_blocks, grid_config.num_threads_per_block, 0, stream>>>
+        (left_valid, right_valid, out_valid, out_col.size(), merged_indices);
+    } else {
+      materialize_merged_bitmask_kernel<true, false>
+        <<<grid_config.num_blocks, grid_config.num_threads_per_block, 0, stream>>>
+        (left_valid, right_valid, out_valid, out_col.size(), merged_indices);
+    }
+  } else {
+    if (p_right_dcol->has_nulls()) {
+      materialize_merged_bitmask_kernel<false, true>
+        <<<grid_config.num_blocks, grid_config.num_threads_per_block, 0, stream>>>
+        (left_valid, right_valid, out_valid, out_col.size(), merged_indices);
+    } else {
+      CUDF_FAIL("materialize_merged_bitmask_kernel<false, false>() should never be called.");
+    }
+  }
+
+  CHECK_STREAM(stream);
+}
+  
+/**
+ * @brief Generates the row indices and source side (left or right) in accordance with the index columns.
+ *
+ *
+ * @tparam index_type Indicates the type to be used to collect index and side information;
+ * @param[in] left_table The left table_view to be merged
+ * @param[in] right_tbale The right table_view to be merged
+ * @param[in] column_order Sort order types of index columns
+ * @param[in] null_precedence Array indicating the order of nulls with respect to non-nulls for the index columns
+ * @param[in] nullable Flag indicating if at least one of the table_view arguments has nulls (defaults to true)
+ * @param[in] stream CUDA stream (defaults to nullptr)
+ *
+ * @Returns A table containing sorted data from left_table and right_table 
+ */
+
+  //BUG: it reverses left-right results
```

Comment:
This comment will break doxygen -- the doxygen block needs to be immediately before the function it documents. 

If there is a bug, we need to fix it and remove this comment. Is there a bug?

Reply:
Fixed.